### PR TITLE
CDAP-658 common-authorization: authorization service and client

### PR DESCRIPTION
**Pull request by jwang47**

### CODE REVIEW

File: common-authorization/src/main/java/co/cask/common/authorization/InMemoryACLStore.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * In-memory implementation of {@link ACLStore}.
+ */
+public class InMemoryACLStore implements ACLStore {
+
+  // (ObjectId, SubjectId) -> set of Permission
+  private final Map<Key, Set<String>> store = Maps.newHashMap();
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/InMemoryACLStore.java]: You can use a Guava SetMultimap here to simplify this

File: common-authorization/src/main/java/co/cask/common/authorization/ObjectId.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+
+/**
+ * Uniquely identifies an object, as defined in {@link ACLStore}.
+ */
+public class ObjectId {
+
+  private final String type;
```
Comment (jwang47) [inline common-authorization/src/main/java/co/cask/common/authorization/ObjectId.java]: I was considering adding a field for `namespace` here, but then users of this library would be forced to use namespace. However, this sort of reasoning also sort of applies to `type`. So, I'm not sure what is best to do here.

File: common-authorization/src/main/java/co/cask/common/authorization/client/AuthorizationClient.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  public boolean setACL(ObjectId objectId, SubjectId subjectId, String permission) throws IOException {
+    String path = String.format("/v1/acls/%s/%s/%s/%s/%s", objectId.getType(), objectId.getId(),
+                                subjectId.getType(), subjectId.getId(), permission);
+    HttpRequest request = HttpRequest.post(resolveURL(path))
+      .withBody(GSON.toJson(new ACLEntry(objectId, subjectId, permission)))
+      .build();
+
+    HttpResponse response = HttpRequests.execute(request);
+    if (response.getResponseCode() == HttpURLConnection.HTTP_OK) {
+      return true;
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/client/AuthorizationClient.java]: What about updating the cache here or clearing cached items?

File: common-authorization/src/main/java/co/cask/common/authorization/client/AuthorizationClient.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+   * @throws IOException if an error occurred when contacting the authorization service
+   */
+  public boolean deleteACL(ObjectId objectId, SubjectId subjectId, String permission) throws IOException {
+    String path = String.format("/v1/acls/%s/%s/%s/%s/%s", objectId.getType(), objectId.getId(),
+                                subjectId.getType(), subjectId.getId(), permission);
+    HttpRequest request = HttpRequest.delete(resolveURL(path)).build();
+
+    HttpResponse response = HttpRequests.execute(request);
+    if (response.getResponseCode() == HttpURLConnection.HTTP_OK) {
+      return true;
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/client/AuthorizationClient.java]: Same as above on updating cache.

File: common-authorization/src/main/java/co/cask/common/authorization/guice/BaseURISupplier.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+package co.cask.common.authorization.guice;
+
+import com.google.common.base.Supplier;
+
+import java.net.URI;
+
+/**
+ * Supplies a {@link URI} to be used as the base URI for interacting with a service.
+ */
+public interface BaseURISupplier extends Supplier<URI> {
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/guice/BaseURISupplier.java]: Why is this needed?  It doesn't add anything to the Supplier interface...

File: common-authorization/src/main/java/co/cask/common/authorization/proxy/CGLibAuthorizationProxyFactory.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  @SuppressWarnings("unchecked cast")
+  public <T> T getProxy(final T object, final ObjectId objectId) {
+    return (T) Enhancer.create(object.getClass(),
+                               new MethodInterceptor() {
+      @Override
+      public Object intercept(Object o, Method method, Object[] args, MethodProxy methodProxy) throws Throwable {
+        RequiresPermissions requiresPermissions = getAnnotation(method.getDeclaredAnnotations());
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/proxy/CGLibAuthorizationProxyFactory.java]: Should this use Method.getAnnotation(RequiresPermissions.class) instead?

File: common-authorization/src/main/java/co/cask/common/authorization/server/AuthorizationHandler.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * Handler for getting and setting ACL entries.
+ */
+@Path("/v1")
+public class AuthorizationHandler extends AbstractHttpHandler {
```
Comment (ghelmling) [inline common-authorization/src/main/java/co/cask/common/authorization/server/AuthorizationHandler.java]: This doesn't seem to have any authentication or authorization itself, so how does it support any authorization enforcement?  What stops a client from setting whatever ACLs they want?  How do we do mutual authentication so that we know we can actually trust the ACLs returned to the client?

**Comment by ghelmling** [comment]
Let's work through the overall architecture of this at a whiteboard.  Between the caching and the client/server communications, I have some questions about how the overall system works and is made performant and reliable.

**Comment by jwang47** [comment]
Made changes according to new design. However, the `ACLManagerService` still needs to be secured.

**Comment by jwang47** [comment]
Going to reopen PR
